gwu-libraries/scholarspace-hyrax
# Add CI workflow and docker-compose healthchecks

This PR:

## Sets up CI workflow using GH Actions

On PR to `master` branch, triggers a workflow to build the docker containers and run the testing suite as a check. If unable to build the docker containers, or any of the RSpec tests fail, the 'check' will fail. If the docker containers build successfully and the RSpec tests pass, the 'check' passes. 

Currently takes about 15 minutes in total to run, though we can likely cut this time down using cacheing of layers in a future iteration. 

## Adds healthchecks to docker-compose 

This PR adds healthchecks in the `docker-compose.yml` file, and reworks the `depends_on` options to work with these healthchecks. This allows each container to start only after the containers that it depends on meet the `service_healthy` condition. We had `depends_on` settings in the docker-compose file, but it didn't seem to work without defining these healthchecks. 

This is helpful for CI/CD, as we can't successfully run the RSpec tests until all of the containers are ready. We can manually check this normally by following logs or retrying, but this should allow us to check those conditions while using GH actions and CI/CD. 

We may want to adjust the specific timing parameters in each healthcheck eventually, but this seems to be working for now. 

## Breakdown of the container order and container dependencies:
- Starts `pg-hyrax`, `pg-fcrepo`, `solr` and `redis` containers, as they don't depend on other containers. 
  - `pg-hyrax` and `pg-fcrepo` use `pg_isready` and their respective DB users and DB names to determine when healthy. 
  - `solr` uses wget to check if the Solr dashboard is accessible.
  - `redis` uses a ping command to determine when healthy. 
- Once `pg-fcrepo` is healthy, `fedora` container starts, and is healthy when wget gets a response from the fedora app landing page. 
- Once `redis` is healthy, `sidekiq` container starts. 
- Once all of `fedora`, `solr`, and `pg-hyrax` are healthy, launches the `app-server` container. 

## To test:
- Build and start docker containers using `docker compose up --wait`
- If successful, containers should start in the order specified, and will start outputting the logs to the terminal only after all of the containers start. 
- Everything else should function the same once all containers are launched, i.e.:
  - Start interactive terminal in the app-server container, e.g. `docker exec -it scholarspace-hyrax_app-server_1 /bin/sh`
  - Change to scholarspace user with `su scholarspace`
  - Migrate prod database with `bundle exec rails db:migrate RAILS_ENV=production`
  - Precompile assets with `bundle exec rails assets:precompile RAILS_ENV=production`
  - Create dev/test databases with `bundle exec rails db:create`
  - Migrate dev/test databases with `bundle exec rails db:migrate`
  - Run `bundle exec rspec` to run tests
